lib/active_record/virtual_attributes/virtual_total.rb

```diff
@@ -72,21 +72,22 @@ def virtual_aggregate(name, relation, method_name = :sum, column = nil, options
       def define_virtual_aggregate_method(name, relation, method_name, column)
         if method_name == :size
           define_method(name) do
-            (attribute_present?(name) ? self[name] : nil) || send(relation).try(:size) || 0
+            (has_attribute?(name) ? self[name] : send(relation).try(:size)) || 0
           end
         else
           define_method(name) do
-            (attribute_present?(name) ? self[name] : nil) ||
-              begin
-                rel = send(relation)
-                if rel.loaded?
-                  rel.blank? ? nil : (rel.map { |t| t.send(column).to_i } || 0).send(method_name)
-                else
-                  # aggregates are not smart enough to handle virtual attributes
-                  arel_column = rel.klass.arel_attribute(column)
-                  rel.try(method_name, arel_column) || 0
-                end
+            if has_attribute?(name)
+              self[name] || 0
+            else
+              rel = send(relation)
+              if rel.loaded?
+                rel.map { |t| t.send(column) }.compact.send(method_name)
```

Comment:
Being blunt:  The commit message for this change is kind of a disaster... for reference:

> ### simplify loaded rel sum
> 
> array.map {} || 0 :
> 
> if there is an array, then it will always have a value,
>   so || 0 is never going to have a value
> 
> if an array is empty, empty.map {} will just be an empty value
> [].sum == 0
> 
> removed to_i since we could sum up floats

```diff
- rel.blank? ? nil : (rel.map { |t| t.send(column).to_i } || 0).send(method_name)
+ rel.map { |t| t.send(column) }.compact.send(method_name)
```


First off, the `compact` was completely unmentioned, but that plays a role in handling `nil` values that existed in the array previously (as `nil.to_i == 0`).

The `[] || 0` is actually a problem (if it was ever reachable) since `0.sum` is not a method on integers.

Finally, you removed `nil` being a valid return value, and that used to not be the case.

Regarding the formatting of the commit message, I think you would be better itemizing the changes (something like this, and probably could be tweaked):

```
simplify loaded rel sum

- removes `|| 0` since `0.sum` is invalid and it is not reachable (since `[] || 0 == []`)
- `[].sum == 0`, so we can avoid needing a `... || 0` case
- removes `nil` as being a possible return value when array is empty (now returns `0`)
- call `.compact` to remove `nil` values fromarray to avoid issue with `.send` (aka: `.sum`) later
```

* * *

The other issue I have is that we are now potentially returning floats when previously we casted to integers, and if that might potentially cause an issue down the line.

Reply:
if the fields are string'd numbers, it worked before and does not work now. but not sure we should work around this. they screwed up.
if the fields were floats or bigints before, it would return a different value (or error) depending upon the source. Now it returns the correct value.

it still returns different values for unloaded vs loaded associations - 0 vs nil - this has not been addressed by this pr. Trying to keep this PR as a documentation/adding tests and less of a change.

Reply:
I changed the comments. will split apart that commit